docs/developer/architecture/centralized-cli.rst

```diff
@@ -149,3 +180,239 @@ Cluster Service Administration
+  to ``stderr``), using
+  `Kubernetes log library <https://github.com/kubernetes/klog>`_.
+
+  By default each command will wait for a result but, when it's possible, a
+  ``--async`` option should allow to do not wait for a result and just output
+  an ID (e.g.: Job ID for Salt) that can be used to watch for the result.
+
+SaltAPI interaction
+"""""""""""""""""""
+
+If the plugin needs to access Salt API then it should use the service proxy
+``http://<apiserver_host>/api/v1/namespaces/kube-system/services/https:salt-master:api/proxy/``.
+
+For each and every Salt API call plugin will need authentication on
+apiserver to access the Salt API service proxy and also to Salt API.
+
+.. note::
+
+  Right now, Salt API only accepts authentication using Bearer token, but in
+  kubeconfig we could have certificates authentication so this kind of
+  kubeconfig will not work with this kubectl plugin.
+
+  Add support for certificates based authentication in Salt API look quite hard
+  and costly.
+
+.. todo::
+
+  Check what can be done to support certificates based kubeconfig or if we
+  consider that we only support token-based kubeconfig for our plugin commands
+  interacting with Salt API.
+
+Deployment
+""""""""""
+
+Plugin should be developed as one single binary ``kubectl-metalk8s`` available
+from the ISO, easy buildable from GitHub repository and also as a System
+Package for Operating System supported by MetalK8s.
+
+The package should install the plugin in ``/usr/bin`` directory by default.
+
+This package should be installed on the bootstrap node by default.
+
+Rejected design choice
+""""""""""""""""""""""
+
+- ``Bash`` kubectl plugin: Bash is great to do simple actions but not when
+  you need to do interaction with some API like Kubernetes API or Salt API.
+- ``Python`` kubectl plugin: Python allows us to do complicated actions and
+  great interactions with APIs but interactions between ``Go`` and Kubernetes
+  are much easier, given the large number of example available.
+
+``metalk8sctl`` CLI
+^^^^^^^^^^^^^^^^^^^
+
+.. _`Salt Python client API`: https://docs.saltstack.com/en/latest/ref/clients/index.html
+
+Language
+""""""""
+
+``Python`` is chosen as the language for ``metalk8sctl`` for the following
+reasons:
+
+- Ability to interact with `Salt Python client API`_.
+- Python installation needed anyway by Salt-minion.
+
+.. note::
+
+  Python version 3 will be used as version 2 is end of life since beginning of
+  2020.
+
+Input and Output
+""""""""""""""""
+
+- Command style:
+
+  .. code-block:: bash
+
+    metalk8sctl <resource> <action>
+
+- Interactive:
+
+  Never.
+
+- Output style:
+
+  Human readable output, do not necessarily need for "machine output" like
+  JSON and YAML.
+
+  The output should display useful information from Salt returns when needed,
+  and in case of error, only show relevant error message(s) from Salt.
+
+.. todo::
+
+  - Define logging informations
```

Comment:
IMO, any kind of *persisted* logging shall be done at the Salt/API/... level (i.e., some kind of audit logs), *not* by some CLI tool/client itself.

Reply:
For `metalk8sctl` IMO we should log stuff in the CLI as we may interact with the machine itself, for the kubectl plugin I agree it's only API driven so it's ok.

Reply:
After bootstrap, is there anything we'd do on the node that's *not* performed through Salt?

Reply:
No you right, but I think we would like to have log to know for example when an upgrade is launched with the result etc, because in Salt logs you will have each Salt call but not a global view.

For me since a command consists of the execution of several API calls (which is not the case for kubectl plugin normally) we want to have a log file for this. (but maybe we can discuss about it)